src/libraries/System.Formats.Nrbf/tests/ArraySinglePrimitiveRecordTests.cs

```diff
@@ -24,6 +26,51 @@ public static IEnumerable<object[]> GetCanReadArrayOfAnySizeArgs()
         }
     }
 
+    [Fact]
+    public void DontCastBytesToBooleans()
+    {
+        using MemoryStream stream = new();
+        BinaryWriter writer = new(stream, Encoding.UTF8);
+
+        WriteSerializedStreamHeader(writer);
+        writer.Write((byte)SerializationRecordType.ArraySinglePrimitive);
+        writer.Write(1); // object ID
+        writer.Write(2); // length
+        writer.Write((byte)PrimitiveType.Boolean); // element type
+        writer.Write((byte)0x01);
+        writer.Write((byte)0x02);
+        writer.Write((byte)SerializationRecordType.MessageEnd);
+        stream.Position = 0;
+
+        SZArrayRecord<bool> serializationRecord = (SZArrayRecord<bool>)NrbfDecoder.Decode(stream);
+
+        bool[] bools = serializationRecord.GetArray();
+        bool a = bools[0];
+        Assert.True(a);
+        bool b = bools[1];
+        Assert.True(b);
+        bool c = a && b;
```

Comment:
I believe you need to use single-ampersand to test what you were trying to test (reliably)

Sharplab shows that sometimes the compiler writes down the IL instruction "and" (which turns into the x86 "and") for both operations, and sometimes it uses brfalse for "&&".

The only time I've seen it use brfalse involved spans, e.g. [this example on sharplab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMCAbJgEYD2dEmAsggBT2OYCGANLQyY0AlJgC8APh6YAZDNoBuYqQzYqArszgdBPfpyGjJ0+TSUpSmZSVUVqBluh1c9Gw9cIfLDgMbiTil4qCACcbD7C5paWMADsmD5RpAC+Hh626g7MACzOTK4OIh5EFtEkvv7csoGlZRRhEUl18YkeqbVWHRn2uswArHkkfG60wsVBJADKAA7cAHYAPAZSBgDO/gDawzQAuk3RE2phaxsADDvVJwg7kYf1eatnF3Juj9e3HZZ38VWm++3JIA==)...

but `&` always ends up as IL:and.

Probably doesn't super matter, as this probably uses IL:and.

Reply:
After finding some boundary examples, I believe that Roslyn turns `expr1 && expr2` into `and` if `expr2` is a literal-, local-, or parameter- expression, and `brfalse` otherwise.  (In my linked example it used `brfalse` because the RHS was an indexer-expression).

If I'm right, then your `&&` here will use the IL `and` instruction, hence the CPU `and` instruction, so it is a valid test.

But, wow, that took a long time to "prove".